src/js_sync.hpp

```diff
@@ -62,11 +64,18 @@ static std::string partition_value_bson_to_string(typename T::Context ctx, typen
         bson::Bson partition_bson;
         if (Value<T>::is_string(ctx, partition_value_value)) {
             std::string pv = Value<T>::validated_to_string(ctx, partition_value_value);
+            if (pv.length() == 0) {
+                throw std::runtime_error("partitionValue of type 'string' cannot be an empty string.");
+            }
             partition_bson = bson::Bson(pv);
         }
         else if (Value<T>::is_number(ctx, partition_value_value)) {
             auto pv = Value<T>::validated_to_number(ctx, partition_value_value);
-            partition_bson = bson::Bson(static_cast<int64_t>(pv));
+            if (pv < 0 || pv > JS_MAX_SAFE_INTEGER || fmod(pv, 1) != 0) {
+                throw std::runtime_error("partitionValue of type 'number' must be a whole number >= 0 and <= Number.MAX_SAFE_INTEGER.");
```

Comment:
I don't think "whole number" is a well-defined mathematical term. Maybe "natural number", "non-negative integer" or just "integer".

Reply:
https://dictionary.cambridge.org/dictionary/english/whole-number 🤷‍♂️
Then I would suggests `"partitionValue of type 'number' must be a non-negative integer <= Number.MAX_SAFE_INTEGER."`?

Reply:
I don't think the `fmod(pv, 1) != 0` guard is safe across compilers and architectures.  It may perform a division (is the user has't done so already), which is not guaranteed to return exact integers, even if the maths say it should.

I think a safer approach would be something like 

```
            auto pv = Value<T>::validated_to_number(ctx, partition_value_value);
            decltype(pv) fraction;
            partition_bson = bson::Bson(static_cast<int64_t>(pv));
            if (pv < 0 || pv > JS_MAX_SAFE_INTEGER || modf(pv, *fraction) > std::numeric_limits<decltype(pv)>::epsilon(); {
               ...
```

Reply:
Thanks @fronck that's just what I needed. I'll test it out & update.

Reply:
@fronck I've updated to the solution we discussed elsewhere.

Reply:
@cesarvr the last "stringification" you mention is into a bson/ejson format, so the type is preserved at that point.

Reply:
Hey @steffenagger that's ok, just wanted to understand that decimal removal validation when the value that comes is a ``number``.  

If you just want only whole numbers you can cast to ``int64_t`` and the number will lost decimal information and you just need to validate for the numerical constraint. 

```cpp
auto number = static_cast<int64_t>(pv)
if (pv < 0 || pv > JS_MAX_SAFE_INTEGER) {
 //throw exception.... 
}
partition_bson = bson::Bson(number);
```

If throwing exception on a decimal is a must then I would suggest: 

```cpp 
  //#include <cmath>
 //  for pv = 10.5;                                                                    
 if (pv < 0 || pv > JS_MAX_SAFE_INTEGER || std::floor(pv) != pv ) { // 10 != 10.5 
  //throw exception.... 
 }
 partition_bson = bson::Bson(pv);
```
I my opinion is simpler way to achieve that goal.

Cheers.

Reply:
Of course, ask away 🙂 (I'll answer to the best of my knowledge).

The exception is a must, as we want to inform the users if they are supplying an unsupported number-format (it would otherwise be completely hidden that we converted the inputted `10.5` to `10`).

The  `int64_t` cast is there to trigger the `Bson` constructor into creating a `Long`  for Sync (not a `Double`).

I do like the simpler `std::floor(pv) != pv` check, @blagoev & @fronck any comments on this?

Reply:
Nice, thanks for the update. 👍